proto/vschema.proto

```diff
@@ -43,6 +44,7 @@ message Keyspace {
   bool sharded = 1;
   map<string, Vindex> vindexes = 2;
   map<string, Table> tables = 3;
+  topodata.KeyspaceType keyspace_type = 4;
```

Comment:
What's the reason for duplicating the field from topopdata.Keyspace? Is it just a way to pass info into vtgate?

Is this proto only used internally, or is this also the basis for the JSON that users provide? Are users ever expected to fill this field, or is it meant to be reserved for internal use (automatically filled from topodata.Keyspace)?

Reply:
Duplicating the field was necessary to pass info into vtgate. This is actually the proto people provide when they create a vschema. In theory people can overwrite the vschema for a snapshot keyspace after first creation. If anyone did that, then they will need to set this field also.

Reply:
If this is the same proto that defines the public API, then this seems like it pollutes the public API to transmit data that the system should be able to know by other means. For example, vtgate could look at the Keyspace objects in topo.

I'm guessing there was some design trade-off that made it ugly for vtgate to look directly at Keyspace objects. Can you elaborate on what that was?

Reply:
We chatted a bit about this out of band and the motivation for providing a timestamp (ideally in addition to a binlog position) was two fold:

1. When we instruct a keyspace to restore to a particular point in time it has the option during bootstraping itself to skip the most recent backup for an older one if the most recent is corrupted in some way. If this happens then we want to have an easy way to expose around what time we would need to start replaying binlogs from since we can no longer assume "most recent backup"
2. Exposing the timestamp is an inexact way to identify when we need to restore data to but for many cases the request comes in to restore at a given time so being able to use that as a marker seems valuable as the alternate approach would be to take the timestamp, scan the available binlogs for a TS=>position mapping then use the discovered binlog position which, effectively, is the same thing

To the comment about the this information should be available elsewhere -- I because the backup discovery process can opt to load the not-most-recent backup it's not something that can know as a given from the keyspace creation request.
As for where that information is exposed: I don't think anybody is particularly attached to the keyspace object. It might make more sense to store this in the `_vt` db or `debug/vars` instead?

Reply:
I like the idea to store metadata about the backup that was used for restore in the `local_metadata` table. That makes more sense to me than putting it in topology.

The question, "what backup was used for restore?" is ultimately a question about the state of data files on disk, not about the conceptual relationship of the tablet to other conceptual entities in topo (keyspaces, shards, other tablets). The info in topo is completely divorced from the reality of what's on disk, but if you wipe the disk to do a restore, you're guaranteed to erase stale `local_metadata` info from a previous restore.

Reply:
I tried to get rid of KeyspaceType from the vschema but haven't been able to find a workable way so far.
Option 1: lookup keyspace type while creating vschema. This doesn't work because vindexes cannot depend on topo
Option 2: pass in topo.Keyspace object to SaveVSchema. Same problem.
Option 3: pass in a bool to specify whether to include in global routing. Problem here is that there is also BuildVSchema which eventually gets down to the same functions (buildTables), and it deals with multiple keyspaces, so we can't pass in a single bool.

If anyone has a better idea, I'm happy to try it.
@enisoc @setassociative

Reply:
Does vtgate always get the vschema from SrvVSchema, and never directly from global topo? If so, I think it would make sense to keep this field, but document that it is only for use in SrvVSchema and will be ignored anywhere else (such as in global topo or in user-submitted vschema definitions).

Then we would fill in `KeyspaceType` from the actual global Keyspace objects during `topo.Server.RebuildSrvVSchema()`. That way, it's impossible for users to cause inconsistency between vschema and the Keyspace record. This feels like a legitimate use of SrvVSchema to me, because in general SrvX records are meant to be cell-local roll-ups of info whose canonical source may be spread across multiple global topo records. In this case, SrvVSchema rolls up some info from VSchema objects and some info from Keyspace objects.

Reply:
Alternatively, we could avoid polluting the `vschema.Keyspace` proto at all if we add a new map of keyspace types to the `SrvVSchema` proto.

Reply:
I tried adding a map of keyspace types to SrvVschema, and then using that in vindexes/vschema.go. We still have the other path from vtctl.go where we have to pass in the KeyspaceType also. With these 2 changes, there is a large number of files that need to change.
```
go/vt/binlog/keyspace_id_resolver.go
go/vt/proto/vschema/vschema.pb.go
go/vt/topo/srv_vschema.go
go/vt/topo/vschema.go
go/vt/vtctl/vtctl.go
go/vt/vtgate/vindexes/vschema.go
go/vt/vtgate/vindexes/vschema_test.go
go/vt/vttablet/tabletserver/vstreamer/engine.go
go/vt/vttablet/tabletserver/vstreamer/planbuilder_test.go
go/vt/worker/legacy_split_clone.go
go/vt/worker/multi_split_diff.go
go/vt/worker/split_clone.go
go/vt/worker/split_diff.go
go/vt/wrangler/migrater.go
proto/vschema.proto
py/vtproto/vschema_pb2.py
```
Diff here: https://gist.github.com/deepthi/13967a9fced4aee085abb90877d07b63
This seems very invasive, so I'm going to try the other suggestion.

Reply:
Coming in late on this one. The high level directive came from me, and was driven by a seemingly harmless requirement: The PITR keyspace should not pollute the routing of existing keyspaces. Because PITR has the same tables as the original keyspace, if we loaded the vschema normally, all those tables will get considered as being present in multiple keyspaces and will be removed from the default serving path.

To make this work, the builder of SrvVSchema has to make sure that a PITR keyspace does not try to add its tables to the global tables list.

There is also a design constraint in vindexes. It's a low level package that must not depend on other infrastructural items, like topo etc. So, adding any topo dependencies there is not a good idea, and I see why deepthi ran into circular dependency issues there.

The other possibility is to have the vschema builder infer this from the keyspace name. The main issue I see with that approach is that we'll have to find a unique naming convention that can in no way conflict with possible naming conventions that a user may use. So, this is what led to the current design.

One way to make this orthogonal to the PITR tablet type is to change the vschema keyspace type to a property that better reflects its functionality: "require explicit routing". This would then add a flexibility to the vschema that will allow users to use this flag for other purposes, and then PITR would just be making use this new vschema feature to prevent table name poisoning.

Thoughts?

Reply:
> One way to make this orthogonal to the PITR tablet type is to change the vschema keyspace type to a property that better reflects its functionality: "require explicit routing". This would then add a flexibility to the vschema that will allow users to use this flag for other purposes, and then PITR would just be making use this new vschema feature to prevent table name poisoning.

That SGTM. My main concern with the original KeyspaceType field was that, being accessible in the user-supplied VSchema, it could become inconsistent with the Keyspace record. If we make the functionality orthogonal as proposed, then each dimension is only specified in one place, so there's no inconsistency.

We still have the sub-optimal UX of, "If you customize the VSchema after creating the snapshot keyspace, you need to make sure to set it to be hidden from global routing." However, if the global routing field in the VSchema becomes part of the official, public VSchema API (not a hidden field for internal use), I feel much better about our ability to mitigate this with documentation.

Reply:
I like this idea. thanks @sougou!